application/models/Token.php

```diff
@@ -214,28 +214,49 @@ public function findByToken($token)
      * Generates a token for this object.
      * @throws CHttpException
      */
-    public function generateToken()
+    public function generateToken($tokenlength = NULL)
     {
-        $iTokenLength = $this->survey->tokenlength;
-        $this->token = $this::generateRandomToken($iTokenLength);
+        $iTokenLength = $tokenlength ? $tokenlength : $this->survey->tokenlength;
+        $this->token = $this->_generateRandomToken($iTokenLength);
+        $this->token = $this->_afterGenerateToken($iTokenLength);
         $counter = 0;
         while (!$this->validate(array('token'))) {
-            $this->token = $this::generateRandomToken($iTokenLength);
+            $this->token = $this->_generateRandomToken($iTokenLength);
+            $this->token = $this->_afterGenerateToken($iTokenLength);
             $counter++;
             // This is extremely unlikely.
-            if ($counter > 10) {
-                throw new CHttpException(500, 'Failed to create unique token in 10 attempts.');
+            if ($counter > 50) {
+                throw new CHttpException(500, 'Failed to create unique token in 50 attempts.');
             }
         }
     }
 
+    /**
+     * Invoke the afterGenerateToken plugin event
+     *
+     * @param integer $iTokenLength
+     * @return string
+     */
+    private function _afterGenerateToken($iTokenLength)
+    {
+        $event = new PluginEvent('afterGenerateToken');
+        $event->set('surveyId', $this->getSurveyId());
+        $event->set('iTokenLength', $iTokenLength);
+        $event->set('oToken', $this);
+        $event->set('token', $this->token);
+        App()->pluginManager->dispatchEvent($event);
+        return $event->get('token');
+    }
+
     /**
      * Creates a random token string without special characters
      *
      * @param integer $iTokenLength
      * @return string
+     *
+     * Be sure to call only from generateToken() to facilitate the pluginevent 'afterGenerateToken'
      */
-    public static function generateRandomToken($iTokenLength)
+    private function _generateRandomToken($iTokenLength)
```

Comment:
Can be static

Reply:
In PHP 7 non-static seems to be faster. See https://www.codedwell.com/post/59/static-vs-non-static-methods-in-php with a timing example for PHP 7.2.6. Besides that it is a bit odd to make a private function static, because it should never be called as $classname::aStaticMethod() from outside the class.

Reply:
On second thoughts: since _afterGenerateToken needs `$this` we might even merge _generateRandomToken and _afterGenerateToken. I will test the third commit locally.

Reply:
Commit 3 proved to be OK (tested in the train from Amsterdam to Rotterdam). In the end we are almost back at the initial idea, except that $_SESSION is avoided and the calling sequence has been streamlined.

Reply:
Would you like the PR to develop as 1 commit or with these 3 commits, so that the history is preserved?

Reply:
About commit, we can squash when we merge anyway, so there will always be one commit no matter how many commits there are in the PR.